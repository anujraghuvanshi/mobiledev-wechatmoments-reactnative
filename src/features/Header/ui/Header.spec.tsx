import renderer, {act} from 'react-test-renderer';
import React from 'react';
import {HeaderComponent, Header} from './Header';
import {IUser, RequestStatus} from '../../../types';
import {Provider} from 'react-redux';
import {store} from '../../../store';

jest.mock('../../../hooks', () => ({
  useAppDispatch: jest.fn(() => jest.fn()),
}));

describe('Header', () => {
  const user: IUser = {
    nick: 'john smith',
    username: 'john',
    avatar: 'avatar.url',
    'profile-image': 'profile-image.url',
  };

  describe('Presentational Component Tests', () => {
    it('should render component', () => {
      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <HeaderComponent user={user} />
          </Provider>,
        );
      });

      const containerElement = component!.root.findByProps({
        testID: 'header-container',
      });
      expect(containerElement).toBeTruthy();
    });

    it('should match component snapshots', () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <HeaderComponent user={user} />
          </Provider>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render the correct name', () => {
      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <HeaderComponent user={user} />
          </Provider>,
        );
      });

      const usernameElement = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(usernameElement.props.children).toBe(user.nick);
    });
  });

  describe('Connected Component and mapStateToProps Tests', () => {
    beforeEach(() => {
      act(() => {
        store.dispatch({
          type: 'users/fetchUser/pending',
          payload: {
            status: RequestStatus.IDLE,
            data: {},
          },
        });
      });
    });

    it('should correctly map user data from state to props', () => {
      store.dispatch({
        type: 'users/fetchUser/fulfilled',
        payload: user,
      });

      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <Header />
          </Provider>,
        );
      });

      const usernameElement = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(usernameElement.props.children).toBe(user.nick);
    });

    it('should update component when store data changes', () => {
      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <Header />
          </Provider>,
        );
      });

      const initialUsername = component!.root.findByProps({
        testID: 'header-username',
      });

      expect(initialUsername.props.children).toBe(undefined);

      const newUser = {
        ...user,
        nick: 'jane doe',
      };

      act(() => {
        store.dispatch({
          type: 'users/fetchUser/fulfilled',
          payload: newUser,
        });
      });

      const updatedUsername = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(updatedUsername.props.children).toBe('jane doe');
    });

    it('should handle empty user data in store', () => {
      act(() => {
        store.dispatch({
          type: 'users/fetchUser/fulfilled',
          payload: {},
        });
      });

      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <Header />
          </Provider>,
        );
      });

      const usernameElement = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(usernameElement.props.children).toBe(undefined);
    });

    it('should handle API Failure store', () => {
      act(() => {
        store.dispatch({
          type: 'users/fetchUser/rejected',
        });
      });

      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <Header />
          </Provider>,
        );
      });

      const usernameElement = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(usernameElement.props.children).toBe(undefined);
    });

    it('should map all required user properties from state', () => {
      act(() => {
        store.dispatch({
          type: 'users/fetchUser/fulfilled',
          payload: user,
        });
      });

      let component;
      act(() => {
        component = renderer.create(
          <Provider store={store}>
            <Header />
          </Provider>,
        );
      });

      const usernameElement = component!.root.findByProps({
        testID: 'header-username',
      });
      expect(usernameElement.props.children).toBe(user.nick);

      const avatarElement = component!.root.findByProps({
        testID: 'user-image',
      });
      expect(avatarElement.props.source.uri).toBe(user.avatar);

      const containerElement = component!.root.findByProps({
        testID: 'header-container',
      });
      const backgroundImage = containerElement.props.children;
      expect(backgroundImage.props.source.uri).toBe(user['profile-image']);
    });
  });
});
