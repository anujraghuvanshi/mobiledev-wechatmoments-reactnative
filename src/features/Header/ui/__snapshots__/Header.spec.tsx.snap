// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Presentational Component Tests should match component snapshots 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "height": 230,
    }
  }
  testID="header-container"
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "#f5f5f5",
        "height": 200,
        "justifyContent": "flex-end",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "profile-image.url",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": 200,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 18,
            "fontWeight": "bold",
            "marginRight": 20,
          }
        }
        testID="header-username"
      >
        john smith
      </Text>
      <Image
        source={
          Object {
            "height": 64,
            "uri": "avatar.url",
            "width": 64,
          }
        }
        style={
          Object {
            "backgroundColor": "#e4f0f5",
            "borderRadius": 8,
            "marginBottom": -24,
            "marginRight": 24,
          }
        }
        testID="user-image"
      />
    </View>
  </View>
</View>
`;
